File: fortran/plot_results.f90
subroutine plot_results(x_surface, y_surface, n_surface, &
                       x_base, y_base, n_base, &
                       grid_x, grid_y, grid_z_surface, grid_z_base, nx, ny, &
                       contour_x, contour_y, n_contour, contour_level)
    ! Save the results data for plotting
    ! input:
    !   x_surface, y_surface     - surface point coordinates
    !   x_base, y_base          - base point coordinates
    !   grid_x, grid_y          - grid coordinates
    !   grid_z_surface, grid_z_base - interpolated elevation data
    !   contour_x, contour_y    - boundary points coordinates
    !   contour_level           - contour level value
    !
    !   openExcav: An Open Source Library for Excavation Calculation
    !   Author(s): Hubery H.B. Woo (hbw8456@163.com)
    !   Copyright 2009-2024 Chongqing Three Gorges University
    
    implicit none
    
    ! Arguments
    integer, intent(in) :: n_surface, n_base, nx, ny, n_contour
    real, intent(in) :: x_surface(n_surface), y_surface(n_surface)
    real, intent(in) :: x_base(n_base), y_base(n_base)
    real, intent(in) :: grid_x(nx,ny), grid_y(nx,ny)
    real, intent(in) :: grid_z_surface(nx,ny), grid_z_base(nx,ny)
    real, intent(in) :: contour_x(n_contour), contour_y(n_contour)
    real, intent(in) :: contour_level
    
    ! Local variables
    integer :: i, j, unit_id
    
    ! Save scatter points data
    open(newunit=unit_id, file='scatter_points.dat', status='replace')
    write(unit_id, *) '# Surface points'
    do i = 1, n_surface
        write(unit_id, *) x_surface(i), y_surface(i)
    end do
    write(unit_id, *) '# Base points'
    do i = 1, n_base
        write(unit_id, *) x_base(i), y_base(i)
    end do
    close(unit_id)
    
    ! Save contour points
    open(newunit=unit_id, file='contour_points.dat', status='replace')
    do i = 1, n_contour
        write(unit_id, *) contour_x(i), contour_y(i)
    end do
    close(unit_id)
    
    ! Save grid surface data
    open(newunit=unit_id, file='grid_surface.dat', status='replace')
    do j = 1, ny
        do i = 1, nx
            write(unit_id, *) grid_x(i,j), grid_y(i,j), grid_z_surface(i,j)
        end do
        write(unit_id, *)  ! Empty line between rows for gnuplot
    end do
    close(unit_id)
    
    ! Save grid base data
    open(newunit=unit_id, file='grid_base.dat', status='replace')
    do j = 1, ny
        do i = 1, nx
            write(unit_id, *) grid_x(i,j), grid_y(i,j), grid_z_base(i,j)
        end do
        write(unit_id, *)  ! Empty line between rows for gnuplot
    end do
    close(unit_id)

end subroutine plot_results

File: matlab/plot_results.m
function plot_results(base, surface, grid, contour_points, params)
    % plot the results
    % input:
    %   base, surface - struct with x, y, z coordinates
    %   grid - struct with X, Y, Z_surface, Z_base
    %   params - struct with contour_level
    %   contour_points - boundary points coordinates
    %
    %   openEarthwork: An Open Source Library for Excavation Calculation
    %   Author(s): Hubery H.B. Woo (hbw8456@163.com)
    %   Copyright 2009-2024 Chongqing Three Gorges University

    % extract the coordinates
    x_surface = surface.x;
    y_surface = surface.y;
    x_base = base.x;
    y_base = base.y;
    X = grid.X;
    Y = grid.Y;
    Z_surface = grid.Z_surface;
    Z_base = grid.Z_base;
    contour_level = params.contour_level;
    
    figure('Position', [100, 100, 1200, 800]);
    
    % scatter points
    subplot(2,2,1)
    plot_points(x_surface, y_surface, x_base, y_base, contour_points);
    
    % surface contour
    subplot(2,2,2)
    plot_surface(X, Y, Z_surface, x_surface, y_surface, contour_points, ...
        contour_level);
    
    % base contour
    subplot(2,2,3)
    plot_base(X, Y, Z_base, x_base, y_base, contour_points, ...
        contour_level);
    
end % function plot_results


function plot_points(x_surface, y_surface, x_base, y_base, contour_points)
    % plot the scatter points
    scatter(x_surface, y_surface, 'filled'); hold on;
    scatter(x_base, y_base, 'filled');
    plot(contour_points(:,1), contour_points(:,2), 'g-', 'LineWidth', 2);
    title('scatter points');
end % function plotScatterPoints


function plot_surface(X, Y, Z_surface, x_surface, y_surface, contour_points, ...
    contour_level)
    % plot the surface contour
    contour(X, Y, Z_surface, contour_level, 'LineWidth', 1.5);
    hold on;
    scatter(x_surface, y_surface, 'filled');
    plot(contour_points(:,1), contour_points(:,2), 'g-', 'LineWidth', 2);
    title('surface contour');
end % function plot_surface

function plot_base(X, Y, Z_base, x_base, y_base, contour_points, ...
    contour_level)
    % plot the base contour
    contour(X, Y, Z_base, contour_level, 'LineWidth', 1.5);
    hold on;
    scatter(x_base, y_base, 'filled');
    plot(contour_points(:,1), contour_points(:,2), 'g-', 'LineWidth', 2);
    title( 'base contour');
end % function plot_base